File: mixed/f1/group1/AddBinary.java
// Given two binary strings a and b, return their sum as a binary string.
// Sample 1:
// Input: a = "1010", b = "1011"
// Output: "10101"

class AddBinary {
    public static void main(String...args) {
        String a = "1010";
        String b = "1011";
        System.out.println("Sum: " + addBinary(a,b));
    }

    static String addBinary(String a, String b) {
        StringBuilder result = new StringBuilder();
        int i = a.length() - 1;
        int j = b.length() - 1;
        int carry = 0;
        while (i >= 0 || j >= 0) {
            int sum = carry;
            if (i >= 0) sum += a.charAt(i--) - '0';
            if (j >= 0) sum += b.charAt(j--) - '0';
            carry = sum > 1 ? 1 : 0;
            result.append(sum % 2);
        }
        if (carry != 0) result.append(carry);

        return result.reverse().toString();
    }
}

File: bytopic/bitmanipulation/python/group1/add_binary.py
"""
Given two binary strings a and b, return their sum as a binary string.

Sample 1:
Input: a = "1010", b = "1011"
Output: "10101"

Tag: 125/150
Tag: 67/2927, R356/2936 (overall frequency ranking)
"""


def add_binary(a: str, b: str) -> str:
    s = []
    carry = 0
    i = len(a) - 1
    j = len(b) - 1
    while i >= 0 or j >= 0 or carry:
        if i >= 0:
            carry += int(a[i])
            i -= 1
        if j >= 0:
            carry += int(b[j])
            j -= 1
        s.append(str(carry % 2))
        carry //= 2
    return ''.join(reversed(s))


def main():
    a = "1010"
    b = "1011"
    print(add_binary(a, b))


if __name__ == "__main__":
    main()
